File: Algorithm/Homework/HW3/min_heapify.py
class MinHeap:
    def __init__(self, arr, length):
        self.arr = arr
        # start at index 1
        # parent => // 2
        # child => 2n or 2n + 1
        self.length = length

    def heapify(self, idx=1):
        if idx > self.length:
            return
        self.heapify(2 * idx)
        self.heapify(2 * idx + 1)

        if

    def printAll(self):
        for i in range(1, self.length + 1):
            print(self.arr[i], end=' ')
        print()

if __name__ == '__main__':
    N = int(input())
    line = (input()).split(' ')
    arr = [int(l) for l in line]
    min_heap = MinHeap(arr=arr, N)
    min_heap.heapify()
    min_heap.printAll()

File: Algorithm/Homework/HW3/min_heapify.cpp
#include <iostream>
using namespace std;

class MinHeap{
private:
    int* arr;
    int length;
public:
    MinHeap(int* arr, int length){
        this->arr = arr;
        // start at index 1
        // parent => / 2
        // child => 2n or 2n + 1
        this->length = length;
    }

    void heapify(int idx){
        int largest = idx;
        if ((2 * idx <= length) && (arr[2 * idx] < arr[largest]))
            largest = 2 * idx;
        if ((2 * idx + 1 <= length) && (arr[2 * idx + 1] < arr[largest]))
            largest = 2 * idx + 1;

        if (largest != idx){
            int temp = arr[idx];
            arr[idx] = arr[largest];
            arr[largest] = temp;
            heapify(largest);
        }
    }

    void printAll(){
        for (int i = 1; i <= length; i++)
            cout << arr[i] << " ";
        cout << endl;
    }
};

int main(){
    // accelerate io time
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    int N;
    cin >> N;
    int arr[N + 1];

    for (int i = 1; i <= N; i++)
        cin >> arr[i];
    MinHeap min_heap = MinHeap(arr, N);

    int heap_i = N / 2;
    while (heap_i >= 1){
        min_heap.heapify(heap_i);
        heap_i--;
    }

    min_heap.printAll();
    return 0;
}
